### ISSUE TITLE
Using library with redirect

### ISSUE THREAD

**Issue by nigeltiany** [dart-lang/oauth2#87]
This library is not friendly to code authorization flow with browser redirect. 
This might have been possible when this issue was first raised https://github.com/dart-lang/oauth2/issues/39 but with the introduction of PKCE, the code verifier will not pass when a new Grant object is created. 

Not an OAuth pro but maybe the library could make the ```_State``` class public with other helper methods such as ```toJson``` and ```fromJson```. It can also contain fields such as code_verifier, code_challenge_method, and state. Clients that cannot open a popup or persist state after redirecting can persist ```_State``` then call ```AuthorizationCodeGrant.FromState``` and resume the code flow

https://github.com/dart-lang/oauth2/blob/d60da4dc391b807a3adb319de6d6493cb36d6760/lib/src/authorization_code_grant.dart#L344

**Comment by lejtmany** [comment]
Probably a better way is to allow the code verifier to be passed in the constructor of AuthorizationCodeGrant and the code verifier can be persisted in localStorage or whatever.

This was mentioned in a [comment](https://github.com/dart-lang/oauth2/pull/59#issuecomment-579673506) by @jonasfj  in the review for the PKCE support but never made it in for some reason.

**Comment by jonasfj** [comment]
@nigeltiany I think a strong argument could made that the `oauth2` package has some poor design choices.. Unless you're doing oauth from a CLI app, it's starts to get a little sketchy...

Maybe it would be better to a new package, that mainly exposes a configuration object with methods, and leaves to the user to handle state serialization, etc. Given some decent guides, I suspect this would be a reasonable approach.

Afaik, there isn't much reason for an oauth to be encapsulating and handling opaque state objects.

--------------

Making this package user friendly to various use-cases will probably be breaking, hence, such efforts are probably best directed towards writing a new package.

**Comment by lejtmany** [comment]
@jonasfj  Wouldn't allowing a `codeVerifier` to be passed into the constructor of `AuthorizationCodeGrant` allow the grant to be recreated properly on refresh if the codeVerifier is persisted (ie. in sessionStorage)? That's the only major hurdle we're running into using this in the browser. Unless there's something else I'm missing?

I can have a PR submitted shortly if that's something you'd consider merging.

**Comment by lejtmany** [comment]
Opened a [PR](https://github.com/dart-lang/oauth2/pull/92)

**Comment by jonasfj** [comment]
@lejtmany, sure that can work... but it's not ergonomic... In fact, for the purpose you have the way objects have to be created and recreated with the same arguments seems absurd.

Hence, I why I would suggest forking to make a new package and cleanup.

**Comment by lazylazyllama** [comment]
Now that we can set the code verifier to restore state, I ask my self how do we get it in the first place. The variable is private, so we can't get read the verifier and save it to localstorage before the page reloads am I correct?

**Comment by jonasfj** [comment]
> I think the codeVerifier should be exposed so that we don't have to generate it ourselves.

I agree, well not really... I agree that it's not ideal that the API does not lend itself to common use-cases.

And I think the best option is to fork this package and spend some time designing a good API interface for an oauth2 package, then create a new package. If you make a good one, I think we could update the readme to recommend that new users consider it.

**Comment by lejtmany** [comment]
> Now that we can set the code verifier to restore state, I ask my self how do we get it in the first place. The variable is private, so we can't read the verifier and save it to localstorage before the page reloads am I correct?
> 
> I think the codeVerifier should be exposed so that we don't have to generate it ourselves.

That's a good point.  I was just assuming the codeVerifier would be generated/persisted and then used to make the grant but I agree that just making the `codeVerifier` public would make things easier.

**Comment by raminsadre** [comment]
The _state must be settable, too. Otherwise, the new grant object will start in _State.initial and you cannot call handleAuthorizationResponse or handleAuthorizationCode (there is a the version of handleAuthorizationCode that doesn't verify the state but it is not public). The clean way to implement this has been described by @nigeltiany, although the q&d way (make everything public) works, too, of course.

**Comment by lejtmany** [comment]
The way I did it was to just call:
```dart
    grant.getAuthorizationUrl(redirectUrl);
```
in the function that handles the redirect.

Admittedly a little kludgy since the authorization url is not actually needed but cleaner than making the state public IMO.

Not really sure how much value that check adds altogether since the fact that an authorization code was received implicitly proves that an auth url was generated. This really only will prevent a case where the code that came back is for a *different* grant but in that case we need a way to disambiguate the responses anyway even if an auth url was generated for *both* grants.

**Comment by raminsadre** [comment]
@lejtmany
> The way I did it was to just call.

Ah, that's clever.
